docs/protocol/agents/relayer.mdx

```diff
@@ -1,35 +1,83 @@
+The indexer writes to a local database (RocksDB) as a means of caching, as well as communicating with the Submitter task - the latter periodically polls the database to check if new messages were dispatched.
 
-Relayers can easily configure the messages that they wish to process. Currently, the relayer will support:
+The relayer prioritizes new messages over old ones, under the assumption that old ones are more likely to have been delivered. This means that when (re)started, two pointers are created: a forward one and a backward one, both initialized to the blockchain tip and using the monotonically increasing `nonce` field of messages to iterate. A similar approach is used to index the other events, but some lack a sequence number and can be missed if incorrect RPC responses are received - which happens very often. As such, the message indexing task sends transaction IDs where it found events to the other indexing tasks, guaranteeing no Hyperlane event is missed as long as it occured within the same transaction as the `Mailbox.dispatch()` call.
 
-1. A sender/recipient whitelist.
-2. A sender/recipient blacklist.
-3. The ability to accept [payments on the origin chain](../../reference/messaging/send.mdx#quote-dispatch) for processing a message on the destination chain.
+### The Submitter
+
+Messages go through four stages as part of the Submitter, which are spawned as independent tasks described below.
 
-For convenience, Hyperlane will run an open source and configurable relayer agent, implemented as a rust binary. If you'd like to run your own relayer, we've open sourced the [binary here](https://github.com/hyperlane-xyz/hyperlane-monorepo/tree/main/rust/agents/relayer).
+<table>
+  <thead>
+    <tr>
+      <th>Submitter Task</th>
+      <th>Description</th>
+    </tr>
+  </thead>
+  <tbody>
+    <tr>
+      <td>Message Processor</td>
+      <td>Polls the local database to check if there are any undelivered messages and pushes them to the Prepare step's queue.</td>
+    </tr>
+    <tr>
+      <td>Prepare Task</td>
+      <td>
+        Pops messages from its queue, ensures gas was paid, fetches any metadata and simulates the message delivery transaction. If the simulation succeeds, the message is pushed to the Submit step's queue. Otherwise, the message is pushed back to the Prepare queue.
+      </td>
+    </tr>
+    <tr>
+      <td>Submit Task</td>
+      <td>
+        Pops messages from its queue, and sends the delivery transaction onchain. deliveries are batched whenever possible. If there is no error broadcasting the transaction, the message is pushed to the Confirm step's queue. Otherwise, the message is pushed back to the Prepare queue.
```

Comment:
```suggestion
        Pops messages from its queue, and sends the delivery transaction onchain. Deliveries are batched whenever possible. If there is no error broadcasting the transaction, the message is pushed to the Confirm step's queue. Otherwise, the message is pushed back to the Prepare queue.
```